pkg/apis/core/v1alpha1/cluster_types.go

```diff
@@ -79,12 +79,58 @@ type KubernetesClusterConnection struct {
 	// The name of the kubeconfig context to use.
 	//
 	// If not specified, will use the default context in the kubeconfig.
+	//
+	// +optional
 	Context string `json:"context,omitempty" protobuf:"bytes,1,opt,name=context"`
 
 	// The default namespace to use.
 	//
 	// If not specified, will use the namespace in the kubeconfig.
+	//
+	// +optional
 	Namespace string `json:"namespace,omitempty" protobuf:"bytes,2,opt,name=namespace"`
+
+	// DefaultRegistryOptions determines where images for this Cluster should
+	// be pushed/pulled from if the Cluster itself does not provide local
+	// registry hosting metadata.
+	//
+	// If not specified, no registry rewriting will occur, and the images will
+	// be pushed/pulled to from the registry specified by the corresponding
+	// image build directive (e.g. `docker_build` or `custom_build`).
+	//
+	// +optional
+	DefaultRegistryOptions *DefaultRegistryOptions `json:"defaultRegistryOptions,omitempty" protobuf:"bytes,3,opt,name=defaultRegistryOptions"`
+}
+
+type DefaultRegistryOptions struct {
```

Comment:
i think we probably want to use a single RegistryHosting type, and add SameName to it? i don't see any real data-modelling reason why they would be different, do you?

Reply:
I think safe to use `RegistryHosting` directly (w/ addition of `SingleName`), and that also opens up some future opportunities to configure a local registry **including the different host{fromClusterNetwork,fromContainerRuntime} values** so makes sense! Done in 2f4a0b5e5